lib/matplotlib/colors.py

```diff
@@ -992,6 +993,169 @@ def scaled(self):
+
+        Examples
+        --------
+        Note this example is equivalent to the `.DivergingNorm` example.
+        >>> import matplotlib.colors as mcolors
+        >>> offset = mcolors.PiecewiseLinearNorm([-2, 0, 4])
+        >>> data = [-2., -1., 0., 1., 2., 3., 4.]
+        >>> offset(data)
+        array([0., 0.25, 0.5, 0.625, 0.75, 0.875, 1.0])
+        """
+        self._data_points = (np.asarray(data_points)
+                                if data_points is not None
+                                else None)
+        self._norm_points = (np.asarray(norm_points)
+                                if norm_points is not None
+                                else None)
+        # self._norm_points = np.asarray(norm_points)
+        self.vmin = np.ma.min(data_points)
+        self.vmax = np.ma.max(data_points)
+
+    def __call__(self, value, clip=None):
+        """
+        Map value to the interval [0, 1]. The clip argument is unused.
+        """
+
+        result, is_scalar = self.process_value(value)
+        self.autoscale_None(result)
+        vmin, vmax = self._data_points[0], self._data_points[-1]
+
+
+        # its possible some of the data points are less than vmin
+        # or vmax.  We could error out if some stop points are
+        # overrun, but this lets things keep going...
+
+        if not np.all(np.diff(self._data_points) >= 0):
+             raise ValueError("data points must increase monotonically")
+        ind = np.where((self._data_points >= vmin) &
+                       (self._data_points <= vmax))[0]
+        result = np.ma.masked_array(np.interp(result,
+                                              self._data_points[ind],
+                                              self._norm_points[ind]),
+                                    mask=np.ma.getmask(result))
+        if is_scalar:
+            result = np.atleast_1d(result)[0]
+        return result
+
+    def autoscale_None(self, A):
+        """
+        Parameters:
+        -----------
+
+        A : tuple
+            data used for autoscaling, if appropriate.
+
+        If ``norm.vmin` is None, or ``norm.vmax`` are none, use
+        the min and max of ``A`` for the endpoints of the linear mapping.
+        """
+
+        # allow autoscaling with the data if the user specifies
+        # vmin or vmax is None.  Note we never set vmax/vmin in the class
+        # except at init.
+        vmin = self.vmin
+        if vmin is None:
+            vmin = np.ma.min(A)
+        vmax = self.vmax
+        if vmax is None:
+            vmax = np.ma.max(A)
+
+        if vmin > vmax:
+            raise ValueError('vmin must be less than vmax')
+
+        if self._data_points is None:
+            self._data_points = np.asarray([vmin, vmax])
+        if self._norm_points is None:
+            N = len(self._data_points)
+            self._norm_points = np.linspace(0, 1, N)
+
+        self._data_points[0] = vmin
```

Comment:
This would modify the input to `PiecewiseLinearNorm` if it was already an array; I don't know if this is something to be concerned about.

Reply:
Now I check that `data_points` is strictly ascending...

Reply:
I think `__init__` should indeed make a copy of the input array to avoid it being invalidated later, and here to avoid changes in place.

Reply:
I am not clear on what you mean here, or why the proposed code is a problem.

Reply:
If someone does
```
blah = [...]
norm = PiecewiseLinearNorm(blah)
# adds the norm to some plot
# use blah for other stuff
```
then `norm` may change `blah` in place, and this will leak into other places that use `blah` (for example, it could be using it in a second, separate instance of PiecewiseLinearNorm).

So the constructor should instead do `self._data_points = np.array(data_points)` which will (unlike `np.asarray`) always perform a copy of data_points.

See #13390 for similar problems related to lack-of-copy.

Reply:
Oh, thanks   I see.  I thought asarray copied...